components/kyma-environment-broker/internal/process/new_provisioning/staged_manager.go

```diff
@@ -0,0 +1,241 @@
+	for _, s := range m.stages {
+		if s.name == stageName {
+			s.AddStep(step, cnd)
+			return nil
+		}
+	}
+	return fmt.Errorf("stage %s not defined", stageName)
+}
+
+func (m *StagedManager) GetAllStages() []string {
+	var all []string
+	for _, s := range m.stages {
+		all = append(all, s.name)
+	}
+	return all
+}
+
+func (m *StagedManager) Execute(operationID string) (time.Duration, error) {
+	operation, err := m.operationStorage.GetOperationByID(operationID)
+	if err != nil {
+		m.log.Errorf("Cannot fetch operation from storage: %s", err)
+		return 3 * time.Second, nil
+	}
+
+	logOperation := m.log.WithFields(logrus.Fields{"operation": operationID, "instanceID": operation.InstanceID, "planID": operation.ProvisioningParameters.PlanID})
+	logOperation.Infof("Start process operation steps for GlobalAcocunt=%s, ", operation.ProvisioningParameters.ErsContext.GlobalAccountID)
+	if time.Since(operation.CreatedAt) > m.operationTimeout {
+		timeoutErr := kebError.TimeoutError("operation has reached the time limit")
+		operation.LastError = timeoutErr
+		defer m.callPubSubOutsideSteps(operation, timeoutErr)
+
+		logOperation.Infof("operation has reached the time limit: operation was created at: %s", operation.CreatedAt)
+		operation.State = domain.Failed
+		_, err = m.operationStorage.UpdateOperation(*operation)
+		if err != nil {
+			logOperation.Infof("Unable to save operation with finished the provisioning process")
+			timeoutErr = timeoutErr.SetMessage(fmt.Sprintf("%s and %s", timeoutErr.Error(), err.Error()))
+			operation.LastError = timeoutErr
+			return time.Second, timeoutErr
+		}
+
+		return 0, timeoutErr
+	}
+
+	var when time.Duration
+	processedOperation := *operation
+
+	for _, stage := range m.stages {
+		if processedOperation.IsStageFinished(stage.name) {
+			continue
+		}
+
+		for _, step := range stage.steps {
+			logStep := logOperation.WithField("step", step.Name()).
+				WithField("stage", stage.name)
+			if step.condition != nil && !step.condition(processedOperation) {
+				logStep.Debugf("Skipping")
+				continue
+			}
+
+			processedOperation, when, err = m.runStep(step, processedOperation, logStep)
+			if err != nil {
+				logStep.Errorf("Process operation failed: %s", err)
+				return 0, err
+			}
+			if processedOperation.State == domain.Failed || processedOperation.State == domain.Succeeded {
+				logStep.Infof("Operation %q got status %s. Process finished.", operation.ID, processedOperation.State)
+				return 0, nil
+			}
+
+			// the step needs a retry
+			if when > 0 {
+				return when, nil
+			}
+		}
+
+		processedOperation, err = m.saveFinishedStage(processedOperation, stage, logOperation)
+		if err != nil {
+			return time.Second, nil
+		}
+	}
+
+	logOperation.Infof("Operation succeeded")
+
+	processedOperation.State = domain.Succeeded
+	m.publisher.Publish(context.TODO(), process.ProvisioningSucceeded{
+		Operation: internal.ProvisioningOperation{Operation: processedOperation},
```

Comment:
we need to send general event like OperationSucceeded, then, in handlers, we need to recognize it by field `type`

Reply:
Pushed.